File: day9b/Program.fs
open System;
open System.IO;
open FSharp.Collections;
open System.Linq;

type Vector = int * int // x, y

let none = Vector(0, 0)
let up = Vector(0, 1)
let down = Vector(0, -1)
let left = Vector(-1, 0)
let right = Vector(1, 0)

let upLeft = Vector(-1, 1)
let upRight = Vector(1, 1)
let downLeft = Vector(-1, -1)
let downRight = Vector(1, -1)


let add (a: Vector) (b: Vector): Vector =
  let (x1, y1) = a
  let (x2, y2) = b
  (x1 + x2, y1 + y2)


let delta (a: Vector) (b: Vector): Vector =
  let (x1, y1) = a
  let (x2, y2) = b
  (x1 - x2, y1 - y2)


type Move = char * int // direction, count


let moveHead (head: Vector) (direction: char): Vector =
  match direction with
  | 'R' -> add head right
  | 'L' -> add head left
  | 'U' -> add head up
  | 'D' -> add head down
  | _ -> failwith $"unexpected direction, \"{direction}\"" 


let getTailMove (head: Vector) (tail: Vector): Vector =
  let diff = delta head tail
  let diffX, diffY = diff

  if diffX >= -1 && diffX <= 1 && diffY >= -1 && diffY <= 1 then
    none
  else if diff = (2, 0) then
    right
  else if diff = (-2, 0) then
    left
  else if diff = (0, 2) then
    up
  else if diff = (0, -2) then
    down
  else if diff = (2, 2) || diff = (2, 1) || diff = (1, 2) then
    upRight
  else if diff = (-2, 2) || diff = (-2, 1) || diff = (-1, 2) then
    upLeft
  else if diff = (2, -2) || diff = (2, -1) || diff = (1, -2) then
    downRight
  else if diff = (-2, -2) || diff = (-2, -1) || diff = (-1, -2) then
    downLeft
  else
    failwith $"getTailMove failed head: {head}, tail: {tail}"


let rec applyMove (head: Vector) (tail: Vector) (move: Move) (tailMoveSet: Set<Vector>): Vector * Vector * Set<Vector> =
  let direction, count = move
  let tailMoveSet = tailMoveSet.Add(tail)

  if count <= 0 then
    head, tail, tailMoveSet

  else
    let head: Vector = moveHead head direction
    let tailMove = getTailMove head tail
    let tail = add tail tailMove

    let remMove = (direction, count - 1)
    applyMove head tail remMove tailMoveSet


let rec apply (head: Vector) (tail: Vector) (moves: List<Move>) (tailMoveSet: Set<Vector>): Set<Vector> =
  let tailMoveSet = tailMoveSet.Add(tail)

  if moves.IsEmpty then
    tailMoveSet
  else
    let firstMove :: otherMoves = moves
    let head, tail, tailMoveSet = applyMove head tail firstMove tailMoveSet
    apply head tail otherMoves tailMoveSet


let parseLine (line: string): Move =
  let tokens = line.Split(" ")
  let direction = tokens[0][0]
  let count = Int32.Parse(tokens[1])
  (direction, count)


let moves = Seq.toList (File.ReadLines("./input.txt").Select(parseLine))
let head = Vector(0, 0)
let tail = Vector(0, 0)
let s: Set<Vector> = Set.empty
let tailMoveSet = apply head tail moves s
printfn $"{tailMoveSet.Count} "

File: day9b/Program.cs
record struct Vector(int X, int Y);
record Move(string Direction, int Count);

class Program
{
    HashSet<Vector> tailPositions = new HashSet<Vector>()
    {
        new Vector(0, 0),
    };
    List<Vector> rope = new List<Vector>();

    static Vector up = new Vector(0, 1);
    static Vector down = new Vector(0, -1);
    static Vector left = new Vector(-1, 0);
    static Vector right = new Vector(1, 0);

    static Dictionary<string, Vector> directionToVector = new Dictionary<string, Vector>()
    {
        { "U", up },
        { "D", down },
        { "L", left },
        { "R", right },
    };

    Vector Add(Vector a, Vector b)
    {
        return new Vector(a.X + b.X, a.Y + b.Y);
    }

    Vector Diff(Vector a, Vector b)
    {
        return new Vector(a.X - b.X, a.Y - b.Y);
    }

    double Amplitude(Vector a)
    {
        return Math.Sqrt(Math.Pow(a.X, 2) + Math.Pow(a.Y, 2));
    }

    void MoveHead(string direction)
    {
        var v = directionToVector[direction];
        rope[0] = Add(rope[0], v);
    }

    Vector GetTailMove(Vector head, Vector tail)
    {
        var delta = Diff(head, tail);
        
        var deltaAmplitude = Amplitude(delta);
        if (deltaAmplitude < 2)
        {
            return new Vector(0, 0);
        }
        
        if (delta == new Vector(2, 0))
        {
            return right;
        }
        else if (delta == new Vector(-2, 0))
        {
            return left;
        }
        else if (delta == new Vector(0, 2))
        {
            return up;
        }
        else if (delta == new Vector(0, -2))
        {
            return down;
        }
        else if (delta == new Vector(2, 2) || delta == new Vector(2, 1) || delta == new Vector(1, 2))
        {
            return new Vector(1, 1);
        }
        else if (delta == new Vector(-2, 2) || delta == new Vector(-2, 1) || delta == new Vector(-1, 2))
        {
            return new Vector(-1, 1);
        }
        else if (delta == new Vector(2, -2) || delta == new Vector(2, -1) || delta == new Vector(1, -2))
        {
            return new Vector(1, -1);
        }
        else if (delta == new Vector(-2, -2) || delta == new Vector(-2, -1) || delta == new Vector(-1, -2))
        {
            return new Vector(-1, -1);
        }
        else 
        {
            throw new Exception($"GetTailMove failed, head={head}, tail={tail}");
        }
    }

    void ApplyMove(string direction)
    {
        MoveHead(direction);

        for (int i = 1; i < rope.Count; i++)
        {
            var move = GetTailMove(rope[i - 1], rope[i]);
            rope[i] = Add(rope[i], move);
        }

        tailPositions.Add(rope.Last());
    }

    void Run()
    {
        var moves = File.ReadLines("./input.txt").Select(line => {
            var tokens = line.Split();
            return new Move(tokens[0], int.Parse(tokens[1]));
        });

        rope = Enumerable.Range(1, 10)
            .Select(_ => new Vector(0, 0))
            .ToList();

        foreach (var move in moves)
        {
            var (direction, count) = move;
            for (var i = 0; i < count; i++)
            {
                ApplyMove(direction);
            }
        }

        Console.Out.WriteLine($"{tailPositions.Count}");
    }

    public static void Main(string[] args)
    {
        var p = new Program();
        p.Run();
    }
}

